OpenRA.Mods.Common/Traits/AttackMove.cs

```diff
@@ -130,16 +141,34 @@ public override void Tick(World world)
 
 		public override string GetCursor(World world, CPos cell, int2 worldPixel, MouseInput mi)
 		{
-			var prefix = mi.Modifiers.HasModifier(Modifiers.Ctrl) ? "assaultmove" : "attackmove";
+			var prefix = mi.Modifiers.HasModifier(Modifiers.Ctrl);
 
-			if (world.Map.Contains(cell) && subjects.Any())
+			var subject = subjects.FirstOrDefault();
+			if (subject != null)
 			{
-				var explored = subjects.First().Actor.Owner.Shroud.IsExplored(cell);
-				var blocked = !explored && subjects.Any(a => !a.Trait.Info.MoveIntoShroud);
-				return blocked ? prefix + "-blocked" : prefix;
+				var info = subject.Trait.Info;
+				if (world.Map.Contains(cell))
```

Comment:
The logic in this block doesn't look correct. We take the cursor from the first actor, but this actor may not be able to move into the shroud.

Reply:
Can you explain that further? I just changed the `Any()` for a null check.

Reply:
I think the idea here is that the blocked cursor of the actor preventing us from moving into shroud should be used instead of the cursor of the first actor. (This was present before, but not visible since people couldn't set the cursor per unit.)

Reply:
Sorry I still don't get it. The logic didn't change. I just replaced hardcoded strings/prefixes with variables. `subjects.First()` was checked against before.

Reply:
The logic *did* change. The existing code doesn't use `First` for `MoveIntoShroud`, but `Any`.

The effect of this can easily be seen ingame: the current behaviour will always show the blocked cursor variant if you have a mix of ground units and aircraft in TS. The PR behaviour will change depending on whether you built the ground unit or aircraft first (which is what determines `subjects.First`). There is also a crash if you activate attack-move while the cursor is over the shroud.

Reply:
The problem is that `subjects` is not a list of actors, but a list of `TraitPair`s. `TraitPair` is a value type, so is never null. The `subject != null` and `shroudAvoider != null` checks are therefore *always* true. The first check can change to `subject.Actor != null` and the second can change to `var cannotMove = subjects.FirstOrDefault(a => !a.Trait.Info.MoveIntoShroud).Trait` (also changing `shroudAvoider`, which is an odd variable name).

Reply:
I changed things, but I am not sure if I got it right.